### DISCUSSION TITLE
Throwing an error inside a queryFn is delayed

### DISCUSSION THREAD

**Discussion by tbowmo** [TanStack/query#5708] [General] [answered]
I'm testing react-query, and in the process, I have the following hook that I call:

```
export function useInviteId() {
    const { id } = useParams<{ id: string }>()

    const query = useQuery({
        queryKey:['inviteId', id],
        queryFn: async () => {
            if (id === 'next') {
                console.log('next')
                return nextInvite()
            }
            const nummericId = Number(id)
            if (isNaN(nummericId)) {
                throw new Error('Not a number')
            }
            return nummericId
        }
    })

    return query.isSuccess ? query.data : -1
}
```

It's using react-router, to get parameters from url (`useParams`), and then queries backend if the id is the word 'next', otherwise it take the id and return it. 

If it's a non nummeric number, I throw an error. However, this part (throwing the error) takes several seconds to propagate out of the queryFunction. All other paths in my queryFn is returning instantly.

So do I have to do anything special to throw the error, and have it returned instantly?

**Comment by tbowmo** [comment]
A minimized reproduction of the code: 

```
function usePromiseRejectDelayed(id: string) {
    return useQuery({
       queryKey: ['somefunction', id],
       queryFunction: () =>  {
          if (id === 'next') {
             return Promise.resolve(1)
          }
         return Promise.reject(new Error('this is delayed')
      }
   })
}
```

If above hook is called with the parameter "next" it returns immediately, but anything else is delayed a few seconds. Tried debugging with query devtools, but as I'm new to  query, I'm on noob grounds as to where to look..

**Comment by TkDodo** [comment, marked as answer]
> However, this part (throwing the error) takes several seconds to propagate out of the queryFunction.

React Query does 3 retries with exponential backoff internally. Set `retry: 0` and you'll see it instantly.

**Reply by tbowmo** [reply]
Where is that documented? (Been all over the documentation trying to figure out what was happening)

**Reply by TkDodo** [reply]
https://tanstack.com/query/v4/docs/react/guides/query-retries
